### PR TITLE
Rework connections UI

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
And related changes

### CODE REVIEW

File: src/R/Wpf/Impl/Converters.cs [hunk 1]
```diff
@@ -28,6 +28,7 @@ public static class Converters {
         public static IValueConverter TrueIsNotHidden { get; } = LambdaConverter.Create<bool>(x => x ? Visibility.Visible : Visibility.Hidden);
         public static IValueConverter FalseIsHidden { get; } = LambdaConverter.Create<bool>(x => !x ? Visibility.Hidden : Visibility.Visible);
         public static IValueConverter FalseIsNotHidden { get; } = LambdaConverter.Create<bool>(x => !x ? Visibility.Visible : Visibility.Hidden);
+        public static IValueConverter FalseIsGrayedOut { get; } = LambdaConverter.Create<bool>(x => x ? 1.0 : 0.25);
```
Comment (AlexanderSher) [inline src/R/Wpf/Impl/Converters.cs]: Is it for transparency

File: src/Common/Core/Impl/Extensions/IOExtensions.cs [hunk 2]
```diff
@@ -60,5 +60,11 @@ public static class IOExtensions {
             return files;
         }
 
+        public static string TrimTrailingSlash(this string path) {
+            if (path.EndsWithOrdinal("\\")) {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/IOExtensions.cs]: We should probably handle both types of slashes:
```{csharp}
if (!string.IsNullOrEmpty(path) && (path.EndsWithOrdinal("\\") || path.EndsWithOrdinal("/"))) {
    return path.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
}
return path;
```

File: src/Host/Client/Impl/Session/RSessionProvider.cs [hunk 3]
```diff
@@ -102,10 +103,9 @@ public Task<RHost> ConnectAsync(string name, IRCallbacks callbacks, string rComm
         }
 
         public Task<bool> TrySwitchBroker(string name, string path = null) {
-            path = path ?? new RInstallation().GetRInstallPath();
-
+            path = path ?? new RInstallation().GetCompatibleEngines().FirstOrDefault()?.InstallPath;
             Uri uri;
-            if (!Uri.TryCreate(path, UriKind.Absolute, out uri)) {
+            if (string.IsNullOrEmpty(path) || !Uri.TryCreate(path, UriKind.Absolute, out uri)) {
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: You don't need this verification - `Uri.TryCreate` handles both nulls and empty strings

File: src/R/Components/Impl/ConnectionManager/Implementation/LocalConnections.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.R.Components.ConnectionManager.Implementation {
+    /// <summary>
+    /// Provides information on locally installed interpreters
+    /// </summary>
+    internal sealed class LocalConnections : IEnumerable<IConnection> {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/LocalConnections.cs]: Do you use this type anywhere?

File: src/R/Components/Impl/ConnectionManager/IConnection.cs [hunk 5]
```diff
@@ -2,15 +2,14 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
 using System;
-using System.Threading.Tasks;
 
 namespace Microsoft.R.Components.ConnectionManager {
-    public interface IConnection {
+    public interface IConnection: IConnectionInfo {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/IConnection.cs]: Where do you use that `IConnection` is derived from  `IConnectionInfo`?

File: src/Package/Impl/Options/R/Tools/RToolsSettingsImplementation.cs [hunk 6]
```diff
@@ -125,14 +123,13 @@ internal sealed class RToolsSettingsImplementation : IRToolsSettings {
...
-        public RToolsSettingsImplementation(IRInteractiveWorkflowProvider workflowProvider) {
+        public RToolsSettingsImplementation() {
             _workingDirectory = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
-            _workflowProvider = workflowProvider;
         }
 
         private async Task SetMirrorToSession() {
-            var sessions = _workflowProvider.GetOrCreate().RSessions.GetSessions();
+            var provider = VsAppShell.Current.ExportProvider.GetExportedValue<IRInteractiveWorkflowProvider>();
+            var sessions = provider.GetOrCreate().RSessions.GetSessions();
```
Comment (AlexanderSher) [inline src/Package/Impl/Options/R/Tools/RToolsSettingsImplementation.cs]: let's make a helper method for it

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs [hunk 7]
```diff
@@ -103,18 +107,13 @@ internal class ConnectionManager : IConnectionManager {
...
-
-        public async Task ConnectAsync(IConnection connection) {
+        public async Task ConnectAsync(IConnectionInfo connection) {
             var sessions = _sessionProvider.GetSessions().ToList();
             if (sessions.Any()) {
                 await Task.WhenAll(sessions.Select(s => s.StopHostAsync()));
             }
 
-            if (ActiveConnection != null && ActiveConnection.Id != connection.Id) {
+            if (ActiveConnection != null && (!ActiveConnection.Path.EqualsIgnoreCase(connection.Path)  || _sessionProvider.BrokerUri.IsLoopback)) {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: Why do we want to special case loopback?

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs [hunk 8]
```diff
@@ -153,25 +152,71 @@ internal class ConnectionManager : IConnectionManager {
...
+                // No connections, may be first use or connections were somehow removed.
+                // Add local connections so there is at least something available.
+                foreach (var e in localEngines) {
+                    var c = CreateConnection(e.Name, e.InstallPath, string.Empty, isUserCreated: false);
+                    connections[new Uri(e.InstallPath, UriKind.Absolute)] = c;
+                }
+            } else {
+                // Remove missing engines and add engines missing from saved connections
+                // Set 'is used created' to false if path points to locally found interpreter
+                foreach (var kvp in connections.ToList()) {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: `connections.Where(c => !c.IsRemote).ToList()`

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs [hunk 9]
```diff
@@ -153,25 +152,71 @@ internal class ConnectionManager : IConnectionManager {
...
+                        }
+                        if (!valid) {
+                            connections.Remove(kvp.Key);
+                        }
+                    }
+                }
+
+                // Add newly installed engines
+                foreach (var e in localEngines) {
+                    if (!connections.Values.Any(x => x.Path.TrimTrailingSlash().EqualsIgnoreCase(e.InstallPath.TrimTrailingSlash()))) {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: Why `TryGetValue` doesn't work?

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml.cs [hunk 10]
```diff
@@ -60,5 +61,21 @@ public partial class ConnectionManagerControl : UserControl {
...
+            }
+        }
+
+        private void Connection_MouseDoubleClick(object sender, MouseButtonEventArgs e) {
+            Model?.ConnectAsync(GetConnection(e)).DoNotWait();
+            e.Handled = true;
+        }
+
+        private bool SuppressClose(object sender) {
+            var cvm = (sender as ListBoxItem)?.DataContext as IConnectionViewModel;
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml.cs]: >(sender as ListBoxItem)?.DataContext as IConnectionViewModel; [](start = 22, length = 61)

GetConnection(e)

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml.cs [hunk 11]
```diff
@@ -60,5 +61,21 @@ public partial class ConnectionManagerControl : UserControl {
...
+                Model?.ConnectAsync(GetConnection(e)).DoNotWait();
+            }
+        }
+
+        private void Connection_MouseDoubleClick(object sender, MouseButtonEventArgs e) {
+            Model?.ConnectAsync(GetConnection(e)).DoNotWait();
+            e.Handled = true;
+        }
+
+        private bool SuppressClose(object sender) {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml.cs]: Is it used?

File: src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionViewModel.cs [hunk 12]
```diff
@@ -24,8 +28,9 @@ internal sealed class ConnectionViewModel : BindableBase, IConnectionViewModel {
             UpdateCalculated();
         }
 
-        public ConnectionViewModel(IConnection connection) {
+        public ConnectionViewModel(IConnection connection, IConnectionManagerViewModel cmvm) {
             _connection = connection;
+            _cmvm = cmvm;
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionViewModel.cs]: Field isn't used

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml [hunk 13]
```diff
@@ -126,116 +125,65 @@
...
+                                   Visibility="{Binding Path=IsRemote, Converter={x:Static rwpf:Converters.FalseIsCollapsed}}"/>
                         <TextBlock Grid.Column="4" Margin="0,1,3,2" Text="{Binding Path=Path, Mode=OneWay}"
-                                   TextTrimming="CharacterEllipsis" TextWrapping="NoWrap"/>
+                                   TextTrimming="CharacterEllipsis" TextWrapping="NoWrap"
+                                   Visibility="{Binding Path=IsRemote, Converter={x:Static rwpf:Converters.FalseIsCollapsed}}"/>
+
+                        <!-- connect button -->
+                        <Button x:Name="ButtonConnect" Grid.Column="5" Style="{StaticResource ToolWindowButtonStyle}" Click="ButtonConnect_Click" 
+                                Visibility="{Binding Path=IsConnected, Converter={x:Static rwpf:Converters.TrueIsCollapsed}}"
+                                Padding="0,0,8,0">
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml]: Is it intentional that `ButtonConnect` is separated from the other 2 buttons?

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml [hunk 14]
```diff
@@ -126,116 +125,65 @@
...
+                        <!-- edit/properties button -->
+                        <Button x:Name="ButtonEdit" Grid.Column="6" Style="{StaticResource ToolWindowButtonStyle}" Click="ButtonEdit_Click"
+                                ToolTip="{x:Static components:Resources.ConnectionManager_EditTooltip}">
+                            <Image Source="{x:Static rwpf:ImageSources.Property}"/>
                         </Button>
 
                         <!-- delete button -->
-                        <Button x:Name="ButtonDelete" Grid.Column="6" Style="{StaticResource ToolWindowButtonStyle}" Click="ButtonDelete_Click"
-                                ToolTip="{x:Static components:Resources.ConnectionManager_DeleteTooltip}" Visibility="Collapsed">
+                        <Button x:Name="ButtonDelete" Grid.Column="7" Style="{StaticResource ToolWindowButtonStyle}" Click="ButtonDelete_Click"
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml]: Shouldn't we just hide this button?

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml [hunk 15]
```diff
@@ -126,116 +125,65 @@
...
-                        <!-- Collapsed -->
-                        <Rectangle Fill="{StaticResource IconArrowRigth16}" Height="16" Width="16" />
-
-                        <!-- Connect button -->
+                        <!-- Connection icon -->
                         <ContentPresenter DataContext="{Binding}" Grid.Column="1" 
                                           ContentTemplate="{StaticResource ConnectButtonDataTemplate}" />
 
-                        <!-- Name & path -->
+                        <!-- Name, path, command line  -->
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml]: I don't see command line field

File: src/R/Wpf/Impl/Converters.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/R/Wpf/Impl/Converters.cs]: Fade icon out so it looks disabled

File: src/Common/Core/Impl/Extensions/IOExtensions.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Extensions/IOExtensions.cs]: OK

File: src/R/Components/Impl/ConnectionManager/Implementation/LocalConnections.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/LocalConnections.cs]: Looks like accidental add. Removed.

File: src/R/Components/Impl/ConnectionManager/IConnection.cs (hunk 5 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/IConnection.cs]: To reuse implementation `ConnectionInfo` and also `IConnection` looks like live connection while `info`  is it's persistent part. Was there a reason they were separate?

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml (hunk 13 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml]: Yes. Less probability of accidental clicks on other buttons.

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml.cs (hunk 11 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml.cs]: Not anymore

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 7 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: This fixes the issue when user had zero connections and no local R. Allows to switch away from the null/stub connector (`NullRHostConnector`). I am open to alternatives, I want to improve the class anyway so it throw more descriptive exception than just `RHostDisconnectedException`

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml (hunk 14 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml]: I considered it, but UI looks a bit odd with gaps on some entries. I'd like to gather some feedback.

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 9 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: Check should be ignoring case and trailing slash existence. Isn't `TryGetValue` using default type comparison for values?

File: src/Package/Impl/Options/R/Tools/RToolsSettingsImplementation.cs (hunk 6 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Options/R/Tools/RToolsSettingsImplementation.cs]: Done

File: src/Package/Impl/Repl/VsRInteractiveWorkflowProvider.cs [hunk 16]
```diff
@@ -65,6 +65,8 @@ internal class VsRInteractiveWorkflowProvider : IRInteractiveWorkflowProvider, I
             return _instanceLazy.Value;
         }
 
+        public IRInteractiveWorkflow Current => (_instanceLazy != null && _instanceLazy.IsValueCreated) ? _instanceLazy.Value : null;
```
Comment (AlexanderSher) [inline src/Package/Impl/Repl/VsRInteractiveWorkflowProvider.cs]: Let's call it `Active`

File: src/R/Components/Impl/ConnectionManager/IConnection.cs (hunk 5 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/IConnection.cs]: They had different fields originally

---
In reply to: [77669938](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77669938) [](ancestors = 77669938)

File: src/Package/Impl/Options/R/Tools/RToolsSettingsImplementation.cs [hunk 17]
```diff
@@ -148,10 +144,10 @@ internal sealed class RToolsSettingsImplementation : IRToolsSettings {
         }
 
         private async Task SetSessionCodePage() {
-            var sessions = _workflowProvider.GetOrCreate().RSessions.GetSessions();
+            var sessions = GetRSessions();
             var cp = RToolsSettings.Current.RCodePage;
- 
-            foreach (var s in sessions.Where(s => s.IsHostRunning)) {
+
+            foreach (var s in GetRSessions().Where(s => s.IsHostRunning)) {
```
Comment (AlexanderSher) [inline src/Package/Impl/Options/R/Tools/RToolsSettingsImplementation.cs]: we already have list of sessions

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 7 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: ok, I've added a `Name` to the `IRHostConnector` in my changes, so I will change this code after merge

---
In reply to: [77671347](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77671347) [](ancestors = 77671347)

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs [hunk 18]
```diff
@@ -55,16 +57,24 @@ internal class ConnectionManager : IConnectionManager {
...
             _userConnections = new ConcurrentDictionary<Uri, IConnection>(userConnections);
 
             UpdateRecentConnections();
             SwitchBrokerToLastConnection();
         }
 
+        private void OnIdle(object sender, EventArgs e) {
+            _shell.Idle -= OnIdle;
+
+            _disposableBag.Add(_statusBar.AddItem(new ConnectionStatusBar {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: This should be part of the `VsStatusBar` implementation. It already subscribes for `_dteEvents.OnStartupComplete`, so we may add a list of pending items that will be added to the status bar when it is ready.

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml (hunk 13 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml]: Then let's use `Margin`, cause `Padding` may not offset the background, which is also considered part of the button

---
In reply to: [77670185](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77670185) [](ancestors = 77670185)

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 9 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: We can add a comparison method to the `IOExtensions` and then change dictionary comparer:
```{csharp}
public static CompareDirectoryPaths(this string path, string otherPath) {
    var pathLength = GetTrailingSlashTrimmedPathLength(path);
    var otherPathLength = GetTrailingSlashTrimmedPathLength(otherPath);
      
    return pathLength == otherPathLength && (pathLength == 0 || string.Compare (path, 0, otherPath, 0, otherPathLength, true, CultureInfo.InvariantCulture) == 0);
}

private static int GetTrailingSlashTrimmedPathLength(string path) {
    if (string.IsNullOrEmpty(path)) {
        return 0;
    } 
    if (path.EndsWithOrdinal("\\") || path.EndsWithOrdinal("/"))) {
        return path.Length - 1;
    }
    return path.Length;
}
```

---
In reply to: [77673244](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77673244) [](ancestors = 77673244)

File: src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml (hunk 14 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/View/ConnectionManagerControl.xaml]: Gaps? That's weird. Ok, let's leave it as it is

---
In reply to: [77672665](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77672665) [](ancestors = 77672665)

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 18 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: It is, the problem was with `DispatchOnUIThread` that causes message pump. I can plain call `AddItem` if you think ctor is only called on the main thread.

Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: If the problem is in the `DispatchOnUIThread`  implementation, shouldn't we fix it in the `VsAppShell`? Maybe as a separate task/bug.

---
In reply to: [77699578](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77699578) [](ancestors = 77699578)

Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: `DispatchOnUIThread` works as expected really, similar to `Dispatcher`. It is safe in 99% but caller does need to be aware that this is effectively a `PostMessage`

Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: ok

---
In reply to: [77704011](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77704011) [](ancestors = 77704011)

File: src/Common/Core/Impl/Extensions/StringExtensions.cs [hunk 19]
```diff
@@ -25,6 +25,9 @@ public static class StringExtensions {
         public static bool EndsWithOrdinal(this string s, string suffix) {
             return s.EndsWith(suffix, StringComparison.Ordinal);
         }
+        public static bool EndsWithOrdinal(this string s, char ch) {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/StringExtensions.cs]: Just `EndsWith` - we don't care about culture for a single char

File: src/Host/Client/Impl/Host/NullRHostConnector.cs [hunk 20]
```diff
@@ -2,13 +2,16 @@
...
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Common.Core;
 
 namespace Microsoft.R.Host.Client.Host {
     internal sealed class NullRHostConnector : IRHostConnector {
-        private static Task<RHost> Result { get; } = TaskUtilities.CreateCanceled<RHost>(new RHostDisconnectedException());
+        private static Task<RHost> Result { get; } = TaskUtilities.CreateCanceled<RHost>(
+            new RHostDisconnectedException(
+                string.Format(CultureInfo.InvariantCulture, Resources.NoConnectionsAvailable, Environment.NewLine, Environment.NewLine)));
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Host/NullRHostConnector.cs]: Resources doesn't support line breaks?

Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Host/NullRHostConnector.cs]: Nope. When submitting test to MessageBox have to do this. `\n` comes as `\n' literal

Comment (AlexanderSher) [inline src/Host/Client/Impl/Host/NullRHostConnector.cs]: We have `xml:space="preserve"` - so it should just work if you type the message on several lines.

---
In reply to: [77705742](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77705742) [](ancestors = 77705742)

Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Host/NullRHostConnector.cs]: Lemme give it a shot. Didn't work well when I needed to space message in the ShowMessage

Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Host/NullRHostConnector.cs]: Ah, the problem is that it is resource editor that doesn't allow formatted text.

**Comment by AlexanderSher** [comment]
:+1:

**Pull request merged**